source/gateway.go

```diff
@@ -0,0 +1,549 @@
+	gateway "sigs.k8s.io/gateway-api/pkg/client/clientset/gateway/versioned"
+	informers "sigs.k8s.io/gateway-api/pkg/client/informers/gateway/externalversions"
+	informers_v1a2 "sigs.k8s.io/gateway-api/pkg/client/informers/gateway/externalversions/apis/v1alpha2"
+
+	"sigs.k8s.io/external-dns/endpoint"
+)
+
+const (
+	gatewayGroup = "gateway.networking.k8s.io"
+	gatewayKind  = "Gateway"
+)
+
+type gatewayRoute interface {
+	// Object returns the underlying route object to be used by templates.
+	Object() kubeObject
+	// Metadata returns the route's metadata.
+	Metadata() *metav1.ObjectMeta
+	// Hostnames returns the route's specified hostnames.
+	Hostnames() []v1alpha2.Hostname
+	// Protocol returns the route's protocol type.
+	Protocol() v1alpha2.ProtocolType
+	// RouteStatus returns the route's common status.
+	RouteStatus() v1alpha2.RouteStatus
+}
+
+type newGatewayRouteInformerFunc func(informers.SharedInformerFactory) gatewayRouteInformer
+
+type gatewayRouteInformer interface {
+	List(namespace string, selector labels.Selector) ([]gatewayRoute, error)
+	Informer() cache.SharedIndexInformer
+}
+
+func newGatewayInformerFactory(client gateway.Interface, namespace string, labelSelector labels.Selector) informers.SharedInformerFactory {
+	var opts []informers.SharedInformerOption
+	if namespace != "" {
+		opts = append(opts, informers.WithNamespace(namespace))
+	}
+	if labelSelector != nil && !labelSelector.Empty() {
+		lbls := labelSelector.String()
+		opts = append(opts, informers.WithTweakListOptions(func(o *metav1.ListOptions) {
+			o.LabelSelector = lbls
+		}))
+	}
+	return informers.NewSharedInformerFactoryWithOptions(client, 0, opts...)
+}
+
+type gatewayRouteSource struct {
+	gwNamespace string
+	gwLabels    labels.Selector
+	gwInformer  informers_v1a2.GatewayInformer
+
+	rtKind        string
+	rtNamespace   string
+	rtLabels      labels.Selector
+	rtAnnotations labels.Selector
+	rtInformer    gatewayRouteInformer
+
+	nsInformer coreinformers.NamespaceInformer
+
+	fqdnTemplate             *template.Template
+	combineFQDNAnnotation    bool
+	ignoreHostnameAnnotation bool
+}
+
+func newGatewayRouteSource(clients ClientGenerator, config *Config, kind string, newInformerFn newGatewayRouteInformerFunc) (Source, error) {
+	ctx := context.TODO()
+
+	gwLabels, err := getLabelSelector(config.GatewayLabelFilter)
+	if err != nil {
+		return nil, err
+	}
+	rtLabels := config.LabelFilter
+	if rtLabels == nil {
+		rtLabels = labels.Everything()
+	}
+	rtAnnotations, err := getLabelSelector(config.AnnotationFilter)
+	if err != nil {
+		return nil, err
+	}
+	tmpl, err := parseTemplate(config.FQDNTemplate)
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := clients.GatewayClient()
+	if err != nil {
+		return nil, err
+	}
+
+	informerFactory := newGatewayInformerFactory(client, config.GatewayNamespace, gwLabels)
+	gwInformer := informerFactory.Gateway().V1alpha2().Gateways() // TODO: Gateway informer should be shared across gateway sources.
```

Comment:
Are we fine with keeping the non shared across resources informers? Could it cause problems?

Reply:
It shouldn't cause problems. The issue is that each informer instance creates another watch on the apiserver. Sharing wherever possible will reduce network traffic and load on the apiserver.

Informers are not shared across existing sources. The node, service, and pod sources are good examples of this duplication.